File: bindings/nodejs/test_smoke.js
/**
 * Node.js bindings smoke/unit tests (cover critical paths as much as possible).
 *
 * Notes:
 * - Does NOT require a real FRPS (uses a minimal mock server for Login/LoginResp only).
 * - Coverage: connect success, auth failure, protocol error, tunnel start/stop, stats, inject Yamux frame -> onData.
 */

const assert = require('assert');
const { Worker } = require('worker_threads');

const { FRPCClient, TunnelType, ErrorCode, cleanup } = require('./frpc_node');

function startMockFrps(mode) {
  // mode: 'ok' | 'auth_fail' | 'wrong_type' | 'missing_run_id'
  //
  // IMPORTANT: FRPCClient.connect() is a synchronous native call and will block the JS main thread.
  // Run the mock server in a Worker thread, otherwise it cannot respond.
  return new Promise((resolve, reject) => {
    const worker = new Worker(
      `
        const net = require('net');
        const { parentPort, workerData } = require('worker_threads');

        function writeLoginRespAndClose(socket, payloadObj) {
          const respBytes = Buffer.from(JSON.stringify(payloadObj), 'utf8');
          const header = Buffer.alloc(9);
          const t = workerData.mode === 'wrong_type' ? 'X' : '1';
          header.writeUInt8(t.charCodeAt(0), 0);
          header.writeBigInt64BE(BigInt(respBytes.length), 1);
          // Write response and close the connection to allow the server to shutdown cleanly.
          socket.end(Buffer.concat([header, respBytes]));
        }

        const sockets = new Set();

        const server = net.createServer((socket) => {
          sockets.add(socket);
          socket.on('close', () => sockets.delete(socket));
          let buf = Buffer.alloc(0);
          socket.on('data', (chunk) => {
            buf = Buffer.concat([buf, chunk]);
            while (buf.length >= 9) {
              const len = Number(buf.readBigInt64BE(1));
              if (len < 0 || len > 1024 * 1024) {
                socket.destroy();
                return;
              }
              if (buf.length < 9 + len) return;
              buf = buf.slice(9 + len); // consume Login

              let resp;
              if (workerData.mode === 'auth_fail') {
                resp = { version: '0.62.1', error: 'bad token' };
              } else if (workerData.mode === 'missing_run_id') {
                resp = { version: '0.62.1' };
              } else {
                resp = { version: '0.62.1', run_id: 'node_test_run_id' };
              }
              writeLoginRespAndClose(socket, resp);
            }
          });
        });

        server.listen(0, '127.0.0.1', () => {
          parentPort.postMessage({ port: server.address().port });
        });

        parentPort.on('message', (msg) => {
          if (msg && msg.cmd === 'close') {
            for (const s of sockets) {
              try { s.destroy(); } catch (_) {}
            }
            server.close(() => process.exit(0));
          }
        });
      `,
      { eval: true, workerData: { mode } }
    );

    worker.on('message', (m) => {
      if (m && m.port) {
        resolve({
          port: m.port,
          close: () => {
            return new Promise((res) => {
              worker.once('exit', () => res());
              worker.postMessage({ cmd: 'close' });
            });
          }
        });
      }
    });
    worker.on('error', reject);
  });
}

async function main() {
  // Positive: connect success + create/start/stop tunnels + send/stats + inject Yamux frame -> onData.
  {
    const { port, close } = await startMockFrps('ok');
    const client = new FRPCClient('127.0.0.1', port, 'test_token');
    try {
      client.connect();

      const t1 = client.createTunnel(TunnelType.STCP_SERVER, 'node_stcp_server', {
        secretKey: 'node_secret',
        localAddr: '127.0.0.1',
        localPort: 8080
      });
      t1.start();
      t1.stop();
      t1.close();

      let got = null;
      const t2 = client.createTunnel(TunnelType.STCP_VISITOR, 'node_stcp_visitor', {
        secretKey: 'node_secret',
        remoteName: 'remote_server',
        bindAddr: '127.0.0.1',
        bindPort: 9090,
        onData: (buf) => {
          got = Buffer.from(buf);
        }
      });
      t2.start();
      const n = t2.sendData(Buffer.from('hello-from-nodejs'));
      assert(n > 0, 'expected sendData > 0');

      // Inject a Yamux DATA frame: version=0, type=DATA(0), flags=0, stream_id=1, length=payload
      const payload = Buffer.from('inbound-from-nodejs', 'utf8');
      const hdr = Buffer.alloc(12);
      hdr.writeUInt8(0, 0);
      hdr.writeUInt8(0, 1);
      hdr.writeUInt16BE(0, 2);
      hdr.writeUInt32BE(1, 4);
      hdr.writeUInt32BE(payload.length, 8);
      const frame = Buffer.concat([hdr, payload]);
      const consumed = t2.injectYamuxFrame(frame);
      assert(consumed > 0, 'expected injectYamuxFrame > 0');
      assert(got && got.equals(payload), 'expected onData to receive injected payload');

      const st = t2.getStats();
      assert(st && typeof st.bytesReceived === 'bigint', 'expected stats.bytesReceived to be BigInt');
      assert(st.bytesReceived > 0n, 'expected bytesReceived > 0');

      t2.stop();
      t2.close();
    } finally {
      client.close();
      await close();
    }
  }

  // Negative: auth failure should surface as ErrorCode.AUTH (-4).
  {
    const { port, close } = await startMockFrps('auth_fail');
    const client = new FRPCClient('127.0.0.1', port, 'bad_token');
    try {
      let ok = false;
      try {
        client.connect();
      } catch (e) {
        ok = e && e.errorCode === ErrorCode.AUTH;
      }
      assert(ok, 'expected auth failure with ErrorCode.AUTH');
    } finally {
      client.close();
      await close();
    }
  }

  // Negative: protocol errors (wrong response type / missing run_id) should surface as ErrorCode.PROTO (-6).
  {
    const { port, close } = await startMockFrps('wrong_type');
    const client = new FRPCClient('127.0.0.1', port, 'test_token');
    try {
      let ok = false;
      try {
        client.connect();
      } catch (e) {
        ok = e && e.errorCode === ErrorCode.PROTO;
      }
      assert(ok, 'expected proto failure with ErrorCode.PROTO on wrong response type');
    } finally {
      client.close();
      await close();
    }
  }

  {
    const { port, close } = await startMockFrps('missing_run_id');
    const client = new FRPCClient('127.0.0.1', port, 'test_token');
    try {
      let ok = false;
      try {
        client.connect();
      } catch (e) {
        ok = e && e.errorCode === ErrorCode.PROTO;
      }
      assert(ok, 'expected proto failure with ErrorCode.PROTO on missing run_id');
    } finally {
      client.close();
      await close();
    }
  }

  cleanup();
  console.log('nodejs bindings smoke test: OK');
}

main().catch((e) => {
  console.error('nodejs bindings smoke test: FAILED', e);
  process.exit(1);
});

File: bindings/python/test_smoke.py
#!/usr/bin/env python3
"""
Python bindings smoke/unit tests (cover critical paths as much as possible).

Notes:
- Does NOT require a real FRPS (uses a minimal mock server for Login/LoginResp only).
- Coverage: connect success, auth failure, protocol error, tunnel start/stop, stats, inject Yamux frame -> data_callback.
"""

import json
import socket
import struct
import threading
import time

from frpc_python import FRPCClient, TunnelType, ErrorCode, cleanup


def _recv_exact(conn: socket.socket, n: int) -> bytes:
    buf = b""
    while len(buf) < n:
        chunk = conn.recv(n - len(buf))
        if not chunk:
            raise RuntimeError("EOF while reading")
        buf += chunk
    return buf


class MockFrps:
    """A minimal FRPS-like server for Login/LoginResp only."""

    def __init__(self, mode: str):
        # mode: "ok" | "auth_fail" | "wrong_type" | "missing_run_id"
        self._mode = mode
        self._stop = threading.Event()
        self._listener = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self._listener.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        self._listener.bind(("127.0.0.1", 0))
        self._listener.listen(1)
        self.port = self._listener.getsockname()[1]
        self._thread = threading.Thread(target=self._serve, daemon=True)
        self._thread.start()

    def close(self):
        self._stop.set()
        try:
            self._listener.close()
        except OSError:
            pass
        self._thread.join(timeout=1.0)

    def _serve(self):
        try:
            conn, _ = self._listener.accept()
        except OSError:
            return
        with conn:
            conn.settimeout(2.0)
            try:
                msg_type = _recv_exact(conn, 1)
                _ = msg_type  # 'o' (Login)
                length_be = _recv_exact(conn, 8)
                length = struct.unpack(">q", length_be)[0]
                if length < 0 or length > 1024 * 1024:
                    return
                _ = _recv_exact(conn, int(length))

                if self._mode == "auth_fail":
                    resp = {"version": "0.62.1", "error": "bad token"}
                    resp_type = b"1"
                elif self._mode == "wrong_type":
                    resp = {"version": "0.62.1", "run_id": "py_test_run_id"}
                    resp_type = b"X"
                elif self._mode == "missing_run_id":
                    resp = {"version": "0.62.1"}
                    resp_type = b"1"
                else:
                    resp = {"version": "0.62.1", "run_id": "py_test_run_id"}
                    resp_type = b"1"

                resp_bytes = json.dumps(resp).encode("utf-8")
                conn.sendall(resp_type + struct.pack(">q", len(resp_bytes)) + resp_bytes)

                # Keep the TCP connection alive for a short time.
                # Current bindings do not read further frames.
                deadline = time.time() + 1.0
                while not self._stop.is_set() and time.time() < deadline:
                    time.sleep(0.01)
            except Exception:
                return


def main() -> int:
    # Positive: connect success + create/start/stop tunnels + send/stats + inject Yamux frame -> data_callback.
    frps_ok = MockFrps("ok")
    try:
        client = FRPCClient("127.0.0.1", frps_ok.port, "test_token")
        client.connect()

        server_tunnel = client.create_tunnel(
            TunnelType.STCP_SERVER,
            "py_stcp_server",
            secret_key="py_secret",
            local_addr="127.0.0.1",
            local_port=8080,
        )
        server_tunnel.start()
        server_tunnel.stop()

        got = {"data": None}

        def on_data(data: bytes):
            got["data"] = data

        visitor_tunnel = client.create_tunnel(
            TunnelType.STCP_VISITOR,
            "py_stcp_visitor",
            secret_key="py_secret",
            remote_name="remote_server",
            bind_addr="127.0.0.1",
            bind_port=9090,
            data_callback=on_data,
        )
        visitor_tunnel.start()
        sent = visitor_tunnel.send_data(b"hello-from-python")
        if sent <= 0:
            raise RuntimeError("expected send_data > 0")

        # Inject a Yamux DATA frame: version=0, type=DATA(0), flags=0, stream_id=1, length=payload
        payload = b"inbound-from-python"
        hdr = bytearray(12)
        hdr[0] = 0  # version
        hdr[1] = 0  # type=DATA
        hdr[2:4] = (0).to_bytes(2, "big")
        hdr[4:8] = (1).to_bytes(4, "big")  # stream_id
        hdr[8:12] = (len(payload)).to_bytes(4, "big")
        frame = bytes(hdr) + payload
        consumed = visitor_tunnel.inject_yamux_frame(frame)
        if consumed <= 0:
            raise RuntimeError("expected inject_yamux_frame > 0")
        if got["data"] != payload:
            raise RuntimeError("expected data_callback to receive injected payload")

        st = visitor_tunnel.get_stats()
        if st.get("bytes_received", 0) <= 0:
            raise RuntimeError("expected bytes_received > 0 after injection")

        visitor_tunnel.stop()

        client.disconnect()
        client.close()
    finally:
        frps_ok.close()

    # Negative: auth failure should surface as ErrorCode.AUTH.
    frps_bad = MockFrps("auth_fail")
    try:
        bad_client = FRPCClient("127.0.0.1", frps_bad.port, "bad_token")
        try:
            bad_client.connect()
            raise RuntimeError("expected connect() to fail with auth error")
        except Exception as e:
            code = getattr(e, "error_code", None)
            if code != ErrorCode.AUTH:
                raise
        finally:
            bad_client.close()
    finally:
        frps_bad.close()

    # Negative: protocol errors (wrong response type / missing run_id) should surface as ErrorCode.PROTO.
    frps_wrong_type = MockFrps("wrong_type")
    try:
        bad_client = FRPCClient("127.0.0.1", frps_wrong_type.port, "test_token")
        try:
            bad_client.connect()
            raise RuntimeError("expected connect() to fail with proto error")
        except Exception as e:
            code = getattr(e, "error_code", None)
            if code != ErrorCode.PROTO:
                raise
        finally:
            bad_client.close()
    finally:
        frps_wrong_type.close()

    frps_missing = MockFrps("missing_run_id")
    try:
        bad_client = FRPCClient("127.0.0.1", frps_missing.port, "test_token")
        try:
            bad_client.connect()
            raise RuntimeError("expected connect() to fail with proto error (missing run_id)")
        except Exception as e:
            code = getattr(e, "error_code", None)
            if code != ErrorCode.PROTO:
                raise
        finally:
            bad_client.close()
    finally:
        frps_missing.close()

    cleanup()
    print("python bindings smoke test: OK")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())


